optapy-core/src/main/python/test/__init__.py

```diff
@@ -0,0 +1,609 @@
+        """
+        Set the solution to be used for this assertion
+        :param solution: Never None
+        """
+        from org.optaplanner.optapy import PythonSolver  # noqa
+        solution_class = get_class(type(solution))
+        wrapped_solution = PythonSolver.wrapProblem(solution_class, solution)
+        return PythonMultiConstraintAssertion(self.delegate.givenSolution(wrapped_solution))
+
+
+class PythonSingleConstraintAssertion:
+    def __init__(self, delegate):
+        self.delegate = delegate
+
+    @overload
+    def penalizes(self) -> None:
+        """
+        Asserts that the Constraint being tested, given a set of facts, results in any number of penalties.
+
+        Ignores the constraint and match weights: it only asserts the number of matches
+        For example: if there are two matches with weight of 10 each, this assertion will succeed.
+        If there are no matches, it will fail.
+
+        :raises AssertionError: when there are no penalties
+        """
+        ...
+
+    @overload
+    def penalizes(self, times: int) -> None:
+        """
+        Asserts that the Constraint being tested, given a set of facts, results in a given number of penalties.
+
+        Ignores the constraint and match weights: it only asserts the number of matches
+        For example: if there are two matches with weight of 10 each, this assertion will check for 2 matches.
+
+        :param times: the expected number of penalties
+        :raises AssertionError: when the expected penalty is not observed
+        """
+        ...
+
+    @overload
+    def penalizes(self, message: str) -> None:
+        """
+        Asserts that the Constraint being tested, given a set of facts, results in any number of penalties.
+
+        Ignores the constraint and match weights: it only asserts the number of matches
+        For example: if there are two matches with weight of 10 each, this assertion will succeed.
+        If there are no matches, it will fail.
+
+        :param message: sometimes None, description of the scenario being asserted
+
+        :raises AssertionError: when there are no penalties
+        """
+        ...
+
+    @overload
+    def penalizes(self, times: int, message: str) -> None:
+        """
+        Asserts that the Constraint being tested, given a set of facts, results in a given number of penalties.
+
+        Ignores the constraint and match weights: it only asserts the number of matches
+        For example: if there are two matches with weight of 10 each, this assertion will check for 2 matches.
+
+        :param times: the expected number of penalties
+        :param message: sometimes None, description of the scenario being asserted
+
+        :raises AssertionError: when the expected penalty is not observed
+        """
+        ...
+
+    def penalizes(self, times: Union[int, str] = None, message: str = None) -> None:
```

Comment:
Why would `times` ever need to be a `str`?

Reply:
Because `times` might actual be message (i.e. this overload call: https://docs.optaplanner.org/latestFinal/optaplanner-javadoc/org/optaplanner/test/api/score/stream/SingleConstraintAssertion.html#penalizes(java.lang.String) ); this is the delegate for every single overload and thus need to handle each of them